doc/dlt-adaptor-stdin.1.md
% DLT-ADAPTOR-STDIN(1)

# NAME

**dlt-adaptor-stdin** - Forward input from stdin to DLT Daemon

# SYNOPSIS

**dlt-adaptor-stdin** \[**-a** apid\] \[**-c** ctid\] \[**-b**\] \[**-t** timeout\] \[**-v** level\] \[**-h**\]

# DESCRIPTION

This is a small external program for forwarding input from stdin to DLT Daemon.

## OPTIONS

-a

:    Set application ID to apid (default: SINA)

-c

:    Set context ID to ctid (default: SINC)

-b

:    Flush buffered logs before unregistering the application

-t

:    Set timeout when sending messages at exit, in ms (default: 10000 = 10sec)

-v

:    Set verbosity level (default: INFO; accepted values: FATAL ERROR WARN INFO DEBUG VERBOSE)

-h

:    Show help

# EXAMPLES

Forward all dmesg to DLT Daemon without discarding any messages
    **dmesg | dlt-adaptor-stdin -b**

Send DBUS messages to DLT Daemon using the program dbus-monitor
    **dbus-monitor | dlt-adaptor-stdin**

# EXIT STATUS

Non zero is returned in case of failure.

# AUTHOR

Saya Sugiura (ssugiura (at) jp.adit-jv (dot) com)

# COPYRIGHT

Copyright (C) 2019 Advanced Driver Information Technology, Bosch and DENSO. License MPL-2.0: Mozilla Public License version 2.0 http://mozilla.org/MPL/2.0/.

# BUGS

See Github issue: <https://github.com/COVESA/dlt-daemon/issues>

# SEE ALSO

**dlt-daemon(1)**


ReleaseNotes.md
# Diagnostic Log and Trace  - Release Notes

Back to [README.md](../README.md)

## Version

2.18.10 STABLE

## Changes

### 2.18.10
## What's Changed
   * Update ReleaseNotes and version to v2.18.9 by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/468
   * Update status badges on README.md by @michael-methner in https://github.com/COVESA/dlt-daemon/pull/467
   * logstorage: Adds option to write logs in gzip format by @LiquidityC in https://github.com/COVESA/dlt-daemon/pull/442
   * Re-Initialize internal logging in daemon mode only by @lvklevankhanh in https://github.com/COVESA/dlt-daemon/pull/459
   * Remove use of DLT_LOG in signal handler by @michael-methner in https://github.com/COVESA/dlt-daemon/pull/472
   * Update gtest_dlt_daemon_multiple_files_logging.cpp by @LocutusOfBorg in https://github.com/COVESA/dlt-daemon/pull/481
   * Update AUTOSAR standard link by @lucafrance in https://github.com/COVESA/dlt-daemon/pull/480
   * cmake: set version to 2.18.9 by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/478
   * Fix macro code  to use boolean value in while instruction (false) by @michael-methner in https://github.com/COVESA/dlt-daemon/pull/469
   * dlt_user_shared: Add timeout to writev by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/385
   * build: add static lib only if necessary by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/479
   * Update CMakeLists.txt by @LocutusOfBorg in https://github.com/COVESA/dlt-daemon/pull/482
   * watchdog: improve dlt watchdog by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/470
   * log-level-config: add option to configure log levels by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/474
   * gtest_dlt_daemon_gateway: dlt-daemon run without dlt_passive.conf file by @lti9hc in https://github.com/COVESA/dlt-daemon/pull/487
   * dlt_unit_test: Fix and improve quality of unit tests by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/494
   * dlt-qnx-system improvement by @lvklevankhanh in https://github.com/COVESA/dlt-daemon/pull/495
   * gtest: Init submodule and update version by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/497
   * dlt-qnx-system: prevent message loss in high load situations by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/490
   * Fix compile error stringop-truncation with GCC 9.4 by @michael-methner in https://github.com/COVESA/dlt-daemon/pull/499
   * Update README.md by @lvklevankhanh in https://github.com/COVESA/dlt-daemon/pull/503
   * dlt-coverage: Add coverage report generator for dlt by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/501
   * dlt-system: move journal reading to its own thread by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/471
   * Switch from GENIVI to COVESA by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/511
   * Update CMakeLists.txt: set required std version to gnu++14 by @LocutusOfBorg in https://github.com/COVESA/dlt-daemon/pull/504
   * cmake: Policy CMP0115 set to OLD behavior for dlt-daemon with cmake >= 3.20 by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/510
   * fix usage of pthread_cond_timedwait by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/491
   * DLT Upstream for minor release by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/515
   * cmake: disable network trace by @alexmohr in https://github.com/COVESA/dlt-daemon/pull/477
   * doc: add COVESA logo image by @minminlittleshrimp in https://github.com/COVESA/dlt-daemon/pull/516

   **Full Changelog**: https://github.com/COVESA/dlt-daemon/compare/v2.18.9...v2.18.10

### 2.18.9

   * dlt-user: fix crash with certain strings (#463)
   * dlt_multiple_files: remove superfluous mode bits and add header file to header list (#462)
   * Android: Add new feature in Android bp (#461)
   * cmake: remove duplicated option message (#454)
   * house-keeper: remove infinite wait (#438)
   * dlt-logd-converter: Fix getting log level from log msg (#456)
   * dlt-logd-converter: fixes android 12 compilation (#445)
   * logfile: exhance internal dlt logging by introducing size limits (#369)
   * This changes a mispatch from fcb676a7 to install the udp binary correctly. (#449)
   * Installs dlt.conf on android (#446)
   * dlt-connection: add socket timeout (#439)
   * Fix memory leak (#441)
   * Check for negative index in dlt_file_message (#437)
   * dlt-user: fix potential non closed socket in init/free (#435)
   * dlt-convert: Fix memory leak by calling dlt_file_free (#434)
   * dlt-user: Fix crashes in dlt_free during dlt_init (#362)
   * Update contacts and removed mailing lists (#431)
   * Updates for Coding Styles (#425)
   * gateway: Fix Node handling and ECUid checks (#429)
   * filetransfer: fix filesize divisible by blocksize case (#383)
   * client: Fix Get Log Info response conversion method (#422)
   * cmake: network trace enable toggle (#424)
   * dlt-system: Fix buffer overflow detection on 32bit targets (#398)
   * dlt-receive: set host interface and allow multiple udp multicast addresses (#420)
   * Fix for Resource and Memory Leak (#418)
   * dlt_daemon_client: Fix Control Msg ECUId comparison with active Gateway (#414)
   * Avoid memory corruption behind buffer wp in function dlt_getloginfo_conv_ascii_to_id (#411)
   * dlt_common: change output of message for log initialization (#412)
   * internal-logging: Fix issues with file logging (#378)
   * systemd: add support for socket activation via systemd (#401)
   * Update maintainer (#410)
   * dlt_daemon_client: Fix change loglevel of application (#408)
   * dlt_client:Block in connect() (#409)
   * dlt-gateway: Fix crash on invalid ip (#381)
   * Update dlt_for_developers.md (#405)
   * logstorage: Truncate ECUid in Logstorage filter to prevent crash (#402)
   * dlt_common.c: Change default logging_mode (#406)
   * dlt-daemon-connection: Start up even if not all bindings are valid (#380)

src/adaptor/dlt-adaptor-stdin.c
/*
 * SPDX license identifier: MPL-2.0
 *
 * Copyright (C) 2011-2015, BMW AG
 *
 * This file is part of COVESA Project DLT - Diagnostic Log and Trace.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License (MPL), v. 2.0.
 * If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * For further information see http://www.covesa.org/.
 */

/*!
 * \author Alexander Wenzel <alexander.aw.wenzel@bmw.de>
 *
 * \copyright Copyright © 2011-2015 BMW AG. \n
 * License MPL-2.0: Mozilla Public License version 2.0 http://mozilla.org/MPL/2.0/.
 *
 * \file dlt-adaptor-stdin.c
 */

/*******************************************************************************
**                                                                            **
**  SRC-MODULE: dlt-adaptor-stdin.c                                           **
**                                                                            **
**  TARGET    : linux                                                         **
**                                                                            **
**  PROJECT   : DLT                                                           **
**                                                                            **
**  AUTHOR    : Alexander Wenzel Alexander.AW.Wenzel@bmw.de                   **
**              Markus Klein                                                  **
**                                                                            **
**  PURPOSE   :                                                               **
**                                                                            **
**  REMARKS   :                                                               **
**                                                                            **
**  PLATFORM DEPENDANT [yes/no]: yes                                          **
**                                                                            **
**  TO BE CHANGED BY USER [yes/no]: no                                        **
**                                                                            **
*******************************************************************************/

/*******************************************************************************
**                      Author Identity                                       **
********************************************************************************
**                                                                            **
** Initials     Name                       Company                            **
** --------     -------------------------  ---------------------------------- **
**  aw          Alexander Wenzel           BMW                                **
**  mk          Markus Klein               Fraunhofer ESK                     **
*******************************************************************************/

/*******************************************************************************
**                      Revision Control History                              **
*******************************************************************************/

/*
 * $LastChangedRevision: 1670 $
 * $LastChangedDate: 2011-04-08 15:12:06 +0200 (Fr, 08. Apr 2011) $
 * $LastChangedBy$
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "dlt_common.h"
#include "dlt_user.h"
#include "dlt_user_macros.h"

#define MAXSTRLEN 1024

#define PS_DLT_APP_DESC      "stdin adaptor application"
#define PS_DLT_CONTEXT_DESC  "stdin adaptor context"

#define PS_DLT_APP "SINA"
#define PS_DLT_CONTEXT "SINC"

DLT_DECLARE_CONTEXT(mycontext)

int main(int argc, char *argv[])
{
    char str[MAXSTRLEN];
    int opt;

    char apid[DLT_ID_SIZE];
    char ctid[DLT_ID_SIZE];
    char version[255];
    int timeout = -1;
    int verbosity = DLT_LOG_INFO;
    int bflag = 0;

    dlt_set_id(apid, PS_DLT_APP);
    dlt_set_id(ctid, PS_DLT_CONTEXT);

    while ((opt = getopt(argc, argv, "a:c:bht:v:")) != -1)
        switch (opt) {
        case 'a':
        {
            dlt_set_id(apid, optarg);
            break;
        }
        case 'c':
        {
            dlt_set_id(ctid, optarg);
            break;
        }
        case 'b':
        {
            bflag = 1;
            break;
        }
        case 't':
        {
            timeout = atoi(optarg);
            break;
        }
        case 'h':
        {
            dlt_get_version(version, 255);

            printf("Usage: dlt-adaptor-stdin [options]\n");
            printf("Adaptor for forwarding input from stdin to DLT daemon.\n");
            printf("%s \n", version);
            printf("Options:\n");
            printf("  -a apid      - Set application id to apid (default: SINA)\n");
            printf("  -c ctid      - Set context id to ctid (default: SINC)\n");
            printf("  -b           - Flush buffered logs before unregistering app\n");
            printf("  -t timeout   - Set timeout when sending messages at exit, in ms (Default: 10000 = 10sec)\n");
            printf(
                "  -v verbosity level - Set verbosity level (Default: INFO, values: FATAL ERROR WARN INFO DEBUG VERBOSE)\n");
            printf("  -h           - This help\n");
            return 0;
            break;
        }
        case 'v':
        {
            if (!strcmp(optarg, "FATAL")) {
                verbosity = DLT_LOG_FATAL;
                break;
            }
            else if (!strcmp(optarg, "ERROR"))
            {
                verbosity = DLT_LOG_ERROR;
                break;
            }
            else if (!strcmp(optarg, "WARN"))
            {
                verbosity = DLT_LOG_WARN;
                break;
            }
            else if (!strcmp(optarg, "INFO"))
            {
                verbosity = DLT_LOG_INFO;
                break;
            }
            else if (!strcmp(optarg, "DEBUG"))
            {
                verbosity = DLT_LOG_DEBUG;
                break;
            }
            else if (!strcmp(optarg, "VERBOSE"))
            {
                verbosity = DLT_LOG_VERBOSE;
                break;
            }
            else {
                printf(
                    "Wrong verbosity level, setting to INFO. Accepted values are: FATAL ERROR WARN INFO DEBUG VERBOSE\n");
                verbosity = DLT_LOG_INFO;
                break;
            }

            break;
        }
        default: /* '?' */
        {
            fprintf(stderr, "Unknown option '%c'\n", optopt);
            return -1;
        }
        }

    DLT_REGISTER_APP(apid, PS_DLT_APP_DESC);
    DLT_REGISTER_CONTEXT(mycontext, ctid, PS_DLT_CONTEXT_DESC);

    if (timeout > -1)
        dlt_set_resend_timeout_atexit((uint32_t)timeout);

    while (fgets(str, MAXSTRLEN, stdin))
        if (strcmp(str, "") != 0)
            DLT_LOG(mycontext, verbosity, DLT_STRING(str));

    DLT_UNREGISTER_CONTEXT(mycontext);

    if (bflag == 1)
        DLT_UNREGISTER_APP_FLUSH_BUFFERED_LOGS();
    else
        DLT_UNREGISTER_APP();

    return 0;
}
